### DISCUSSION TITLE
Benchmark TorchSharp GPT NEO performance with respect to those of PyTorch and C++ (ONNX)

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [xamarin/TorchSharp#967] [Ideas]
[GPT NEO: better performance of python GPT NEO than its onnx runtime version in C++](https://github.com/microsoft/onnxruntime/issues/15191)

In this issue, it seems PyTorch performs better than the c++ (ONNX runtime).

This is related to TorchSharp Deep NLP or NLG (Natural Language Generation) [discussions](https://github.com/dotnet/TorchSharp/discussions/457#discussioncomment-5369824)

The codes are provided. 

I hope community members here could work together to benchmark Torchsharp using this hot GPT topic given the benchmark codes and use case provided by the community member from ONNX Runtime.

FYI
@fwaris
@ChengYen-Tang

```python
import torch

from transformers import GPTNeoForCausalLM, GPT2Tokenizer
model = GPTNeoForCausalLM.from_pretrained("EleutherAI/gpt-neo-1.3B")
tokenizer = GPT2Tokenizer.from_pretrained("EleutherAI/gpt-neo-1.3B")
device = torch. Device('cpu')
```

- GPT2Tokenizer for .NET is available [here](https://github.com/dotnet/machinelearning/pull/6272)
- [GPT-NEO ](https://huggingface.co/EleutherAI/gpt-neo-1.3B/tree/main) and [WIP instruction/discussion](https://discuss.huggingface.co/t/what-is-rust-model-ot-and-how-can-i-use-it/769) on how to convert PyTorch weight file from HuggingFace to be usable for TorchSharp

**Comment by NiklasGustafsson** [comment]
@GeorgeS2019 -- is this an issue or a discussion topic?

**Comment by GeorgeS2019** [comment]
@NiklasGustafsson 

This issue has been added to [ONNX Runtime backlog](https://github.com/microsoft/onnxruntime/issues/15191#issuecomment-1483381292)

This is a request (most likely long-term) for the benefit for [new users interested to push for TorchSharp GPT](https://github.com/dotnet/TorchSharp/discussions/457#discussioncomment-5370580) to consider a contribution to add an additional selling point for TorchSharp's readiness in production through the "official" benchmarking with respect to PyTorch and ONNX c++

**Comment by GeorgeS2019** [comment]
Similar class of [TorchSharp benchmarking is provided here](https://github.com/zxgamekingdom/PytorchDeploy)

**Comment by NiklasGustafsson** [comment]
Okay, I'm going to convert this to a discussion. This is not a feature request or a bug, or a report of missing features or documentation. If you think benchmarks are important to the future of TorchSharp, please make the case for it in more detail in the discussion thread and build the community support for it there. I would suggest broadening the benchmarking case to include other scenarios than GPT.
